### PR TITLE
Create key if does not already exist

### LINKED ISSUE #3: Creating Key
The biggest issue we need to figure is who creates the key. ARM (acs-engine template) won't create the key. So the grpc server needs to create the key if it does not exist (handling all the race conditions that can happen from multiple create calls).

There are two choices here:

Default acs-engine (we need to make sure that masters have access to CREATE): key created automatically by gRPC server 
Optional (we don't need top support this in first acs-engine phase): User reference an existing keyvault, key and version. in this case we don't need to add CREATE permission

cc @khenidak

### PR DESCRIPTION
**Pull request by ritazh**
Fixes #3 

- Verify provided key vault exists, if not, error
- Get key version of the provided key, if key does not exist, create it
- Verify key version if provided, if key version does not exist, error. If key version is not provided, get key version with the provided key name

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -57,14 +57,14 @@ Mount `/tmp` to access the socket:
...
+    "providerVaultName": "",
     "providerKeyName": "",
     "providerKeyVersion": ""
 
 ```
-* `providerVaultBaseURL`: should look something like this: `https://yourkeyvaultname.vault.azure.net/`
-* `providerKeyName`: name of the key created in azure key vault
-* `providerKeyVersion`: key version of the key created in azure key vault
+* `providerVaultName`: name of the key vault you have created in the same resource group as your k8s cluster. If the provided key vault is not found, service will return an error.
+* `providerKeyName`: name of the key created in azure key vault. If the provided key is not found, service will create a key `providerKeyName` in the `providerVaultName` key vault.
```
Comment (khenidak) [inline README.md:66]: I think this shouldn't be required, if no key, then we create a key using `well known name`  on behalf of the user. this enables bring-up experience to just focus on creating keyvaults and assign the right permissions.

File: oauth.go [hunk 2]
```diff
@@ -61,9 +61,12 @@ type AzureAuthConfig struct {
 // All fields are required unless otherwise specified
 type Config struct {
 	AzureAuthConfig
-
-	// The kms provider vault base url
-	ProviderVaultBaseURL string `json:"providerVaultBaseURL" yaml:"providerVaultBaseURL"`
+	// Location for cluster
```
Comment (khenidak) [inline oauth.go]: do we support both *.yaml and *.json config files? if not then let us focus on one struct tags type.

File: oauth.go [hunk 3]
```diff
@@ -102,20 +105,22 @@ func GetKMSProvider(configFilePath string) (vaultBaseUrl *string, keyName *strin
 			return nil, nil, nil, err
 		}
 		
-		if (config.ProviderVaultBaseURL != "" && config.ProviderKeyName != "" && config.ProviderKeyVersion != "" ) {
-			vaultBaseUrl = &config.ProviderVaultBaseURL
+		if (config.ProviderVaultName != "" ) {
```
Comment (khenidak) [inline oauth.go:106]: what happens if one of these `required` values is missing?

File: oauth.go [hunk 4]
```diff
@@ -143,19 +148,91 @@ func GetKeyvaultToken(grantType OAuthGrantType, configFilePath string) (authoriz
...
 			return nil, err
 		}
 		authorizer = autorest.NewBearerAuthorizer(servicePrincipalToken)
 		return authorizer, nil
-	} else {
-		return
+	} 
+	return
+}
+func GetSubscriptionId(configFilePath string) (subscriptionId *string, err error) {
```
Comment (khenidak) [inline oauth.go]: isn't subscription id part of `azure.json`? just carry `AzureConfig` as part of your data structure. (same for the rest of the fields)

File: server.go [hunk 5]
```diff
@@ -32,39 +36,126 @@ const (
 	runtime        = "Microsoft AzureKMS"
 	runtimeVersion = "0.0.1"
 	configFilePath  = "/etc/kubernetes/azure.json"
+	azureResourceURL = "https://management.azure.com/"
+	vaultResourceURL = "https://vault.azure.net"
```
Comment (khenidak) [inline server.go]: sovereign clouds will carry a different url

File: server.go [hunk 6]
```diff
@@ -32,39 +36,126 @@ const (
 	runtime        = "Microsoft AzureKMS"
 	runtimeVersion = "0.0.1"
 	configFilePath  = "/etc/kubernetes/azure.json"
+	azureResourceURL = "https://management.azure.com/"
```
Comment (khenidak) [inline server.go]: nit: s/URL/Url

File: server.go [hunk 7]
```diff
@@ -32,39 +36,126 @@ const (
...
-	vaultBaseURL, keyName, keyVersion, err := GetKMSProvider(configFilePath)
+	token, err := GetKeyvaultToken(AuthGrantType(), configFilePath, vaultResourceURL)
 	if err != nil {
-		return vmClient, "", "", "", fmt.Errorf("failed to get KMS provider configs, error: %v", err)
+		return kvClient, "", "", "", fmt.Errorf("failed to get token, error: %v", err)
+	}
+	
+	kvClient.Authorizer = token
+	_, keyName, keyVersion, err := GetKMSProvider(configFilePath)
+	if keyName == nil {
```
Comment (khenidak) [inline server.go]: this should be part of `GetKMSProvider`, the code here should do `if nil != err` and shouldn't filter error

File: server.go [hunk 8]
```diff
@@ -32,39 +36,126 @@ const (
...
+		return kvClient, "", "", "", fmt.Errorf("failed to find KMS providerKeyName in config")
+	}
+	if keyVersion == nil {
+		fmt.Println("Unable to find KMS providerKeyVersion in configs, getting keyVersion")
+		keyVersion, err := getKeyVersion(kvClient, *vaultURL, *keyName)
```
Comment (khenidak) [inline server.go]: the keyvault client treats empty `""` keyvault version as `latest`, I think we should assume empty key version on configuration == `stay on latest`

File: server.go [hunk 9]
```diff
@@ -32,39 +36,126 @@ const (
...
+	subscriptionID, _ := GetSubscriptionId(configFilePath)
+	vaultsClient := kvmgmt.NewVaultsClient(*subscriptionID)
+	token, _ := GetKeyvaultToken(AuthGrantType(), configFilePath, azureResourceURL)
+	vaultsClient.Authorizer = token
+	return vaultsClient
+}
+
+func getVault() (vaultURL *string, err error) {
+	vaultsClient := getVaultsClient()
+	providerVaultName, _, _, err := GetKMSProvider(configFilePath)
```
Comment (khenidak) [inline server.go]: we can safety cache the return of `GetKMSProvider`, let us not get it everytime.

**Review by khenidak — CHANGES_REQUESTED** [review]
few changes here and there. But we are very close

File: README.md (hunk 1 above)
Comment (ritazh) [reply README.md:66]: my thought was to allow acs-engine to provide the `default` values for these two parameters in `azure.json`. So the default bring-up experience is the same, but if users have their own keyvault or key, then they can update these two parameters thru acs-engine as well. wdyt?

File: oauth.go (hunk 3 above)
Comment (ritazh) [reply oauth.go:106]: If `ProviderVaultName` or `ProviderKeyName` is missing from the config, then we return an error `failed to find KMS providerKeyName in config`. acs-engine should be providing these parameters in `azure.json`.

File: server.go (hunk 5 above)
Comment (ritazh) [reply server.go]: what's the best place to store this?

File: README.md (hunk 1 above)
Comment (khenidak) [reply README.md:66]: I agree.

File: oauth.go (hunk 3 above)
Comment (khenidak) [reply oauth.go:106]: yeah following last comment this now makes sense

**Review by khenidak — APPROVED** [review]

**Pull request merged**